File: agent-governance-python/agentmesh-integrations/mastra-agentmesh/src/audit.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
/**
 * Audit middleware for Mastra tool execution.
 *
 * Provides tamper-evident audit logging with SHA-256 hash chains.
 * Every tool invocation, completion, denial, and error is recorded
 * with a cryptographic link to the previous entry.
 */

import type { AuditConfig, AuditEntry, GovernanceResult, TrustVerification } from "./types";

let previousHash = "0000000000000000000000000000000000000000000000000000000000000000";
const entries: AuditEntry[] = [];
let entryCounter = 0;

/**
 * Creates an audit middleware that records all tool executions
 * with tamper-evident hash chains.
 *
 * @example
 * ```ts
 * const audit = auditMiddleware({
 *   captureData: true,
 *   maxEntries: 1000,
 *   sink: async (entry) => {
 *     await db.insert("audit_log", entry);
 *   },
 * });
 *
 * const entry = await audit.record({
 *   toolId: "search",
 *   agentId: "agent-1",
 *   action: "invoke",
 *   input: { query: "test" },
 * });
 * ```
 */
export function auditMiddleware(config: AuditConfig = {}) {
  const maxEntries = config.maxEntries ?? 10_000;

  return {
    /**
     * Record an audit entry with hash chain integrity.
     */
    async record(params: {
      toolId: string;
      agentId: string;
      action: "invoke" | "complete" | "deny" | "error";
      input?: unknown;
      output?: unknown;
      duration_ms?: number;
      governance?: GovernanceResult;
      trust?: TrustVerification;
    }): Promise<AuditEntry> {
      const id = `audit-${++entryCounter}-${Date.now()}`;
      const timestamp = Date.now();

      const hashPayload = JSON.stringify({
        id,
        timestamp,
        toolId: params.toolId,
        agentId: params.agentId,
        action: params.action,
        previousHash,
      });

      const hash = await computeHash(hashPayload);

      const entry: AuditEntry = {
        id,
        timestamp,
        toolId: params.toolId,
        agentId: params.agentId,
        action: params.action,
        input: config.captureData ? params.input : undefined,
        output: config.captureData ? params.output : undefined,
        duration_ms: params.duration_ms,
        governance: params.governance,
        trust: params.trust,
        hash,
        previousHash,
      };

      previousHash = hash;
      entries.push(entry);

      // Trim old entries
      while (entries.length > maxEntries) {
        entries.shift();
      }

      // Send to custom sink
      if (config.sink) {
        await config.sink(entry);
      }

      return entry;
    },

    /**
     * Get all audit entries (most recent first).
     */
    getEntries(limit?: number): AuditEntry[] {
      const result = [...entries].reverse();
      return limit ? result.slice(0, limit) : result;
    },

    /**
     * Verify the integrity of the audit chain.
     * Returns true if no entries have been tampered with.
     */
    async verifyChain(): Promise<{ valid: boolean; brokenAt?: number }> {
      for (let i = 1; i < entries.length; i++) {
        if (entries[i].previousHash !== entries[i - 1].hash) {
          return { valid: false, brokenAt: i };
        }
      }
      return { valid: true };
    },

    /**
     * Get entry count.
     */
    get length(): number {
      return entries.length;
    },

    /**
     * Clear all entries (for testing).
     */
    clear() {
      entries.length = 0;
      entryCounter = 0;
      previousHash = "0000000000000000000000000000000000000000000000000000000000000000";
    },
  };
}

export type { AuditEntry };

/** Compute SHA-256 hash of a string. Works in Node.js and Edge. */
async function computeHash(data: string): Promise<string> {
  // Node.js crypto
  try {
    const { createHash } = await import("crypto");
    return createHash("sha256").update(data).digest("hex");
  } catch {
    // Fallback for edge runtimes
    const encoder = new TextEncoder();
    const buf = await crypto.subtle.digest("SHA-256", encoder.encode(data));
    return Array.from(new Uint8Array(buf))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("");
  }
}

File: agent-governance-python/agentmesh-integrations/openai-agents-trust/src/openai_agents_trust/audit.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
"""Tamper-evident audit logging with hash chain."""

from __future__ import annotations

import hashlib
import json
import time
import threading
from dataclasses import dataclass, field
from typing import Dict, List, Optional


@dataclass(frozen=True)
class AuditEntry:
    """An immutable audit log entry."""

    timestamp: float
    agent_id: str
    action: str
    decision: str  # "allow", "deny", "warn"
    details: dict = field(default_factory=dict)
    previous_hash: str = ""
    entry_hash: str = ""

    def to_dict(self) -> dict:
        return {
            "timestamp": self.timestamp,
            "agent_id": self.agent_id,
            "action": self.action,
            "decision": self.decision,
            "details": self.details,
            "previous_hash": self.previous_hash,
            "entry_hash": self.entry_hash,
        }


class AuditLog:
    """Tamper-evident audit log using hash chain hashing."""

    def __init__(self):
        self._entries: List[AuditEntry] = []
        self._lock = threading.Lock()

    def _compute_hash(self, entry_data: dict, previous_hash: str) -> str:
        payload = json.dumps(entry_data, sort_keys=True) + previous_hash
        return hashlib.sha256(payload.encode()).hexdigest()

    def record(
        self,
        agent_id: str,
        action: str,
        decision: str,
        details: Optional[dict] = None,
    ) -> AuditEntry:
        """Record a new audit entry with hash chain hash."""
        with self._lock:
            previous_hash = self._entries[-1].entry_hash if self._entries else ""
            entry_data = {
                "timestamp": time.time(),
                "agent_id": agent_id,
                "action": action,
                "decision": decision,
                "details": details or {},
            }
            entry_hash = self._compute_hash(entry_data, previous_hash)
            entry = AuditEntry(
                timestamp=entry_data["timestamp"],
                agent_id=agent_id,
                action=action,
                decision=decision,
                details=details or {},
                previous_hash=previous_hash,
                entry_hash=entry_hash,
            )
            self._entries.append(entry)
            return entry

    def verify_chain(self) -> bool:
        """Verify the integrity of the entire audit chain."""
        with self._lock:
            for i, entry in enumerate(self._entries):
                previous_hash = self._entries[i - 1].entry_hash if i > 0 else ""
                if entry.previous_hash != previous_hash:
                    return False
                entry_data = {
                    "timestamp": entry.timestamp,
                    "agent_id": entry.agent_id,
                    "action": entry.action,
                    "decision": entry.decision,
                    "details": entry.details,
                }
                expected = self._compute_hash(entry_data, previous_hash)
                if entry.entry_hash != expected:
                    return False
            return True

    def get_entries(
        self,
        agent_id: Optional[str] = None,
        decision: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[AuditEntry]:
        """Query audit entries with optional filters."""
        with self._lock:
            entries = list(self._entries)

        if agent_id:
            entries = [e for e in entries if e.agent_id == agent_id]
        if decision:
            entries = [e for e in entries if e.decision == decision]
        if limit:
            entries = entries[-limit:]
        return entries

    def __len__(self) -> int:
        return len(self._entries)
